File: linked_list/234_palindrome_linked_list.cc
#include <vector>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// 2024/01/07: 自己写出来: 将值复制到数组中后用双指针法
bool IsPalindrome(ListNode* head) {
  std::vector<int> res;
  while (head) {
    res.push_back(head->val);
    head = head->next;
  }
  int left = 0, right = res.size() - 1;
  while (left <= right) {
    if (res[left++] != res[right--]) {
      return false;
    }
  }
  return true;
}

// 2024/01/07: 抄的答案, 递归方法. 不需要记忆.
bool RecursivelyCheck(ListNode* current_node) {
  if (current_node) {
    if (!RecursivelyCheck(current_node->next)) {
      return false;
    }
    if (current_node->val != front_pointer_->val) {
      return false;
    }
    front_pointer_ = front_pointer_->next;
  }
  return true;
}
ListNode* front_pointer_;
bool isPalindrome2(ListNode* head) {
  front_pointer_ = head;
  return RecursivelyCheck(head);
}

// 找到前半部分链表的尾节点
// 反转后半部分链表
// 判断是否回文
// 恢复链表
// 返回结果
bool IsPalindrome3(ListNode* head) {
  if (head == nullptr) {
    return true;
  }

  // 找到前半部分链表的尾节点并反转后半部分链表
  ListNode* first_half_end = EndOfFirstHalf(head);
  ListNode* second_half_start = ReverseList(first_half_end->next);

  // 判断是否回文
  ListNode* p1 = head;
  ListNode* p2 = second_half_start;
  bool result = true;
  while (result && p2 != nullptr) {
    if (p1->val != p2->val) {
      result = false;
    }
    p1 = p1->next;
    p2 = p2->next;
  }

  // 还原链表并返回结果
  first_half_end->next = ReverseList(second_half_start);
  return result;
}

ListNode* ReverseList(ListNode* head) {
  ListNode* prev = nullptr;
  ListNode* curr = head;
  while (curr != nullptr) {
    ListNode* next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  return prev;
}

ListNode* EndOfFirstHalf(ListNode* head) {
  ListNode* fast = head;
  ListNode* slow = head;
  while (fast->next != nullptr && fast->next->next != nullptr) {
    fast = fast->next->next;
    slow = slow->next;
  }
  return slow;
}

File: linked_list/234_palindrome_linked_list.py
import sys, os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from typing import Optional
from list_node import ListNode, create_list_node, print_list_node


# 2026/05/13: 自己实现时间 O(n), 空间 O(n).
def is_palindrome(head: Optional[ListNode]) -> bool:
    nums = []
    curr = head
    while curr:
        nums.append(curr.val)
        curr = curr.next

    for i in range(len(nums) // 2):
        if nums[i] != nums[len(nums) - 1 - i]:
            return False
    return True


# 2026/05/13: 直接看答案
# 通过递归可以反向遍历节点. 时间复杂度 O(n), 由于递归会使用堆栈空间, 所以空间复杂度 O(n).
def is_palindrome2(head: Optional[ListNode]) -> bool:
    front_pointer = head

    def recursively_check(current_node=head) -> bool:
        if current_node is not None:
            if not recursively_check(current_node.next):
                return False
            if front_pointer.val != current_node.val:
                return False
            front_pointer = front_pointer.next
        return True

    return recursively_check()


# 2026/05/13: 直接抄答案.
# 找到前半部分链表的尾节点 -> 反转后半部分链表 -> 判断是否回文 -> 恢复链表 -> 返回结果
def is_palindrome3(head: Optional[ListNode]) -> bool:
    if head is None:
        return True

    first_half_end = end_of_first_half(head)
    second_half_start = reverse_list(first_half_end.next)

    result = True
    first_position = head
    second_position = second_half_start
    while result and second_position is not None:  # 注意这里的结束条件.
        if first_position.val != second_position.val:
            result = False
        first_position = first_position.next
        second_position = second_position.next
    first_half_end.next = reverse_list(second_half_start)
    return result


def end_of_first_half(head: Optional[ListNode]) -> Optional[ListNode]:
    fast, slow = head, head
    while fast.next is not None and fast.next.next is not None:
        fast = fast.next.next
        slow = slow.next
    return slow


def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    prev = None
    curr = head
    while curr:
        next = curr.next
        curr.next = prev
        prev = curr
        curr = next

    return prev


head = [1, 2, 2, 2, 1]
head = create_list_node(head)
print_list_node(head)
print(is_palindrome3(head))
